internal/ethapi/api.go

```diff
@@ -1202,35 +1188,56 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 		log.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
 		hi = gasCap
 	}
-	cap = hi
 
-	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64, state *state.StateDB, header *types.Header) (bool, *core.ExecutionResult, error) {
+	// Create a helper that executes the transaction under a given gas limit and returns true
+	// if the transaction fails for a reason that might be related to not enough gas. A non-nil
+	// error means execution failed due to reasons unrelated to the gas limit.
+	executable := func(gas uint64, state *state.StateDB) (bool, *core.ExecutionResult, error, uint64) {
 		args.Gas = (*hexutil.Uint64)(&gas)
 
-		result, err := doCall(ctx, b, args, state, header, nil, nil, 0, gasCap)
+		result, gasUsed, err := doCall(ctx, b, args, state, header, nil, nil, 0, gasCap)
 		if err != nil {
 			if errors.Is(err, core.ErrIntrinsicGas) {
-				return true, nil, nil // Special case, raise gas limit
+				return true, nil, nil, 0 // Special case, raise gas limit
 			}
-			return true, nil, err // Bail out
+			return true, nil, err, 0 // Bail out
 		}
-		return result.Failed(), result, nil
+		return result.Failed(), result, nil, gasUsed
 	}
-	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
-	if state == nil || err != nil {
+
+	// We first execute the transaction at the highest allowable gas limit, since if this fails we
+	// can return error immediately.
+	failed, result, err, gasUsed := executable(hi, st.Copy())
+	if err != nil {
 		return 0, err
 	}
-	// Execute the binary search and hone in on an executable gas limit
+	if failed {
+		if result != nil && result.Err != vm.ErrOutOfGas {
+			if len(result.Revert()) > 0 {
+				return 0, newRevertError(result)
+			}
+			return 0, result.Err
+		}
+		return 0, fmt.Errorf("gas required exceeds allowance (%d)", hi)
+	}
+	// For almost any transaction, the gas consumed by the unconstrained execution above
+	// lower-bounds the gas limit required for it to succeed. One exception is those txs that
+	// explicitly check gas remaining in order to successfully execute within a given limit, but we
+	// probably don't want to return a lowest possible gas limit for these cases anyway.
+	lo = gasUsed - 1
+
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
```

Comment:
Not really so much binary search if you adjust the mid point yeah?

Reply:
It's a pure binary search after 1-2 doublings of 'lo' in the typical case, lg(n) in the worst. Wasn't sure it's worth clarifying that detail, but I certainly could if it helps.

Reply:
I think it's still basically a binary search, it divides up the range of possible outcomes every iteration, and discards one of the two subranges. The trick here is that the point where the range is divided is not on on the arithmetic middle. 

Some clarifying might be good, e.g. what you wrote below, something like 

> most txs don't need much higher gas limit than their gas used, and most txs don't require near the full block limit of gas, so the selection of where to bisect the range is skewed to favour the low side.

Reply:
added.